cognite/extractorutils/rest/extractor.py

```diff
@@ -354,6 +380,34 @@ def prepare_headers(self, endpoint: Endpoint) -> Dict[str, str]:
 
         return headers
 
+    def _get_next_call(self) -> Optional[HttpEndpointCall]:
+        waiting: Optional[PrioritizedEndpointCall] = None
+        while not self.cancelation_token.is_set():
+            # If n_executing is set to 0, and the queue is empty here, then we are done
+            # This is race condition prone in theory, but it should work. n_executing is
+            # only decremented after giving the call a chance to add to the call queue, so
+            # either the call is unfinished, and n_executing is > 0, or the queue length is
+            # accurate.
+            if self.n_executing == 0 and self.call_queue.empty():
+                return waiting.call if waiting is not None else None
+            to_wait = self._min_check_interval if waiting is None else waiting.call.call_when - time.time()
+            if waiting is not None and to_wait <= 0:
+                return waiting.call
+            try:
+                next: PrioritizedEndpointCall = self.call_queue.get(block=True, timeout=to_wait)
+            except Empty:
+                continue
+
+            if waiting is not None and next.call.call_when < waiting.call.call_when:
+                self.call_queue.put(waiting)
+                waiting = next
+            elif waiting is None:
+                waiting = next
+            else:
+                time.sleep(min(self._min_check_interval, waiting.call.call_when - time.time()))
```

Comment:
Use `self.cancelation_token.wait(...)` instead of `time.sleep(...)`. If the token is set while sleeping, this will block shutdown, while wait with timeout will sleep for the duration unless the token is set in the meantime